producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,293 @@
+		setKeepAliveError: scope.Counter("set-keep-alive-error"),
+	}
+}
+
+type connectFn func(addr string) (net.Conn, error)
+
+type consumerWriterImpl struct {
+	// encodeLock controls the access to the encode function.
+	encodeLock sync.Mutex
+	// decodeLock controls the access to the decode function.
+	decodeLock sync.Mutex
+	encdec     proto.EncodeDecoder
+
+	addr            string
+	retrier         retry.Retrier
+	router          ackRouter
+	opts            Options
+	resetDelayNanos int64
+	logger          log.Logger
+
+	lastResetNanos *atomic.Int64
+	closed         *atomic.Bool
+	doneCh         chan struct{}
+	resetCh        chan struct{}
+	wg             sync.WaitGroup
+	m              consumerWriterMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerWriter(
+	addr string,
+	router ackRouter,
+	opts Options,
+) consumerWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	w := &consumerWriterImpl{
+		encdec:          proto.NewEncodeDecoder(defaultConn, opts.EncodeDecoderOptions()),
+		addr:            addr,
+		retrier:         retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		router:          router,
+		opts:            opts,
+		resetDelayNanos: int64(opts.ConnectionResetDelay()),
+		logger:          opts.InstrumentOptions().Logger(),
+		lastResetNanos:  atomic.NewInt64(0),
+		closed:          atomic.NewBool(false),
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newConsumerWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	w.connectFn = w.connectOnce
+	if err := w.resetWithConnectFn(w.connectFn); err != nil {
+		w.signalReset()
+	}
+	return w
+}
+
+// Write should fail fast so that the write could be tried on other
+// consumer writers that are sharing the data.
+func (w *consumerWriterImpl) Write(m proto.Marshaler) error {
+	w.encodeLock.Lock()
+	err := w.encdec.Encode(m)
+	w.encodeLock.Unlock()
+	if err != nil {
+		w.m.encodeError.Inc(1)
+		w.signalReset()
+	}
+	return err
+}
+
+func (w *consumerWriterImpl) signalReset() {
+	// Avoid resetting too frequent.
+	if w.nowFn().UnixNano() < w.lastResetNanos.Load()+w.resetDelayNanos {
+		return
+	}
+	select {
+	case w.resetCh <- struct{}{}:
+	default:
+	}
+}
+
+func (w *consumerWriterImpl) Init() {
+	w.wg.Add(2)
+	go w.readAcksForever()
+	go w.resetConnectionForever()
+}
+
+func (w *consumerWriterImpl) readAcksForever() {
+	defer w.wg.Done()
+
+	var acks msgpb.Ack
+	for !w.closed.Load() {
+		w.decodeLock.Lock()
+		err := w.encdec.Decode(&acks)
+		w.decodeLock.Unlock()
+		if err != nil {
+			w.m.decodeError.Inc(1)
+			w.signalReset()
+			// Adding some delay here to avoid this being retried in a tight loop
+			// when underlying connection is misbehaving.
+			time.Sleep(w.opts.AckErrorRetryDelay())
```

Comment:
nit: would it be worth using the [`retrier`](https://github.com/m3db/m3x/blob/master/retry/retry.go) package in m3x so we can perform some form of exponential backoff?

Reply:
Yeah I don't exponential backoff is necessary in this case, will keep an eye on it if it turns out we could use some backoff here.